Std/Data/Array/Lemmas.lean

```diff
@@ -844,3 +848,61 @@ instance [DecidableEq α] (a : α) (as : Array α) : Decidable (a ∈ as) :=
 /-! ### erase -/
 
 @[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+
+/-! ### swap -/
+
+@[simp] theorem get_swap_right (a : Array α) {i j : Fin a.size} :
+    (a.swap i j)[j.val]'(a.size_swap .. |>.symm ▸ j.isLt) = a[i] :=
+  by simp [swap, fin_cast_val]
+
+@[simp] theorem get_swap_left (a : Array α) {i j : Fin a.size} :
+    (a.swap i j)[i.val]'(a.size_swap .. |>.symm ▸ i.isLt) = a[j] :=
+  if he : ((Array.size_set _ _ _).symm ▸ j).val = i.val then by
+    simp [←he, fin_cast_val]
+  else by
+    apply Eq.trans
+    · apply Array.get_set_ne
+      · simp
+      · assumption
+    · simp [get_set_ne]
+
+@[simp] theorem get_swap_of_ne (a : Array α) {i j : Fin a.size} (hp : p < a.size)
+    (hi : p ≠ i) (hj : p ≠ j) : (a.swap i j)[p]'(a.size_swap .. |>.symm ▸ hp) = a[p] := by
+  apply Eq.trans
+  · have : ((a.size_set i (a.get j)).symm ▸ j).val = j.val := by simp only [fin_cast_val]
+    apply Array.get_set_ne
+    · simp [this]
```

Comment:
Use `simp?` to turn this non-terminal simp into a `simp only`.

Reply:
done